staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -48,30 +48,38 @@ type Validator struct {
 
 // NewValidator returns compiles all the CEL programs defined in x-kubernetes-validations extensions
 // of the Structural schema and returns a custom resource validator.
-func NewValidator(s *schema.Structural) Validator {
+func NewValidator(s *schema.Structural) *Validator {
 	return validator(s, true)
 }
 
-func validator(s *schema.Structural, isResourceRoot bool) Validator {
+func validator(s *schema.Structural, isResourceRoot bool) *Validator {
 	compiledRules, err := Compile(s, isResourceRoot)
-	result := Validator{compiledRules: compiledRules, compilationErr: err, isResourceRoot: isResourceRoot}
 	if s.Items != nil {
```

Comment:
Properties, and AdditionalProperties are mutually exclusive:
https://github.com/kubernetes/kubernetes/blob/e4c795168befad6a7f372837e6917b5fe8b95d6e/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L748 (cannot have both properties and additional properties in the same schema)

I thought objects vs. arrays were exclusive (because their object types are different), but the're not quite ☹️ . Here's the behavior:

When creating a CRD with a structural schema, it must have one type (unless it is int-or-string or an bare preserve-unknown-fields in which case it can have no type), and type=['object', 'array'] is not allowed (this results in a error creating the CRD). This is documented and I did some black box verification and it seems to hold.

But, I am able to create a a CRD with a type=object schema and with both properties and items set. However, if I try to create a custom resource for it with an array, it get an error that a map is accepted.

I'll go ahead and update the validation code to not make any assumptions. It will just construct validators any of the three branches. This solves for the case where items and properties are set.

The structural schema -> CEL declaration code still checks the object type and decides what type to map to based on that. so it won't do anything with items when type="object".

@sttts Is this okay? Is there a reason why type=object can have an items?

Reply:
> @sttts Is this okay? Is there a reason why type=object can have an items?

probably an oversight in forbidding it... could open an issue to tighten though if it's harmless I suspect it's not worth the effort